packages/compiler/src/server/tspconfig/completion.ts

```diff
@@ -211,3 +354,240 @@ export async function provideTspconfigCompletionItems(
+  const textEdit = getFullCompletionItemValAndEditPosition(
+    newText,
+    source,
+    nodePosRange,
+    curPos,
+    tspConfigPosition,
+  );
+  if (textEdit !== undefined) {
+    return {
+      label: labelName,
+      kind: CompletionItemKind.Field,
+      documentation: description,
+      textEdit: textEdit,
+    };
+  }
+  return undefined;
+}
+
+/**
+ * Find a set of dirs/files with the same suffix
+ * @param rootPath The absolute input path or relative path of the current configuration file
+ * @param fileExtension  File extension, such as ".yaml" or ".tsp"
+ * @param configFile absolute path of the current configuration file
+ * @returns dir/file array
+ */
+async function findFilesOrDirsWithSameExtension(
+  rootPath: string,
+  fileExtension: string,
+  configFile: string = "",
+): Promise<string[]> {
+  const exclude = ["node_modules", "tsp-output", ".vs", ".vscode"];
+
+  const files: string[] = [];
+  try {
+    // When reading the content under the path, an error may be reported if the path is incorrect.
+    // Exclude the current configuration file, compare in absolute paths
+    (await fs.readdir(rootPath, { withFileTypes: true }))
+      .filter(
+        (d) =>
+          (d.isDirectory() || (d.isFile() && d.name.endsWith(fileExtension))) &&
+          !exclude.includes(d.name) &&
+          sysPath.resolve(d.name) !== configFile,
+      )
+      .map((d) => {
+        files.push(d.name);
+      });
+  } catch {
+    return files;
+  }
+  return files;
+}
+
+/**
+ * Get the CompletionItem object array of the relative path of the file
+ * @param relativeFiles File relative path array
+ * @param siblings Sibling node array
+ * @returns CompletionItem object array
+ */
+function getFilePathCompletionItems(relativeFiles: string[], siblings: string[]): CompletionItem[] {
+  const items: CompletionItem[] = [];
+  if (relativeFiles.length > 0) {
+    for (const file of relativeFiles) {
+      if (!siblings.includes(file)) {
+        const item: CompletionItem = {
+          label: file,
+          kind: CompletionItemKind.File,
+        };
+        items.push(item);
+      }
+    }
+  }
+  return items;
+}
+
+/**
+ * resolve variable interpolation completion items
+ * @param target YamlScalarTarget object, contains the information of the target node
+ * @param tspConfigPosition current cursor position in the configuration file
+ * @param targetLine current line content of the target node
+ * @returns variable interpolation completion items
+ */
+function resolveVariableInterpolationCompleteItems(
+  target: YamlScalarTarget,
+  tspConfigPosition: Position,
+  targetLine: string,
+): CompletionItem[] {
+  const yamlDocNodes = target.yamlDoc;
+
+  if (target.sourceType === "QUOTE_DOUBLE" || target.sourceType === "QUOTE_SINGLE") {
+    const targetPos = targetLine.lastIndexOf("{");
+    const curText = targetLine.substring(targetPos, tspConfigPosition.character);
```

Comment:
you should be able to get what you need from source and pos, right? so you dont need to process the doc again.